cpu/atmega_common/periph/rtt.c

```diff
@@ -0,0 +1,249 @@
+    TCCR2B = (1 << CS22) | (1 << CS21) | (1 <<  CS20);
+
+    /* Wait until not busy anymore */
+    DEBUG("RTT waits until ASSR not busy\n");
+    _asynch_wait();
+
+    /* Clear interrupt flags */
+    /* Oddly, this is done by writing ones; see datasheet */
+    TIFR2 = (1 << OCF2B) | (1 << OCF2A) | (1 << TOV2);
+
+    /* Enable 8-bit overflow interrupt */
+    TIMSK2 |= (1 << TOIE2);
+
+    DEBUG("RTT initialized\n");
+}
+
+void rtt_set_overflow_cb(rtt_cb_t cb, void *arg)
+{
+    /* Interrupt safe order of assignment */
+    rtt_state.overflow_cb = NULL;
+    rtt_state.overflow_arg = arg;
+    rtt_state.overflow_cb = cb;
+}
+
+void rtt_clear_overflow_cb(void)
+{
+    /* Interrupt safe order of assignment */
+    rtt_state.overflow_cb = NULL;
+    rtt_state.overflow_arg = NULL;
+}
+
+uint32_t rtt_get_counter(void)
+{
+    /* Make sure it is safe to read TCNT2, in case we just woke up */
+    DEBUG("RTT sleeps until safe to read TCNT2\n");
+    TCCR2A = 0;
+    _asynch_wait();
+
+    return (((uint32_t)rtt_state.ext_cnt << 8) | (uint32_t)TCNT2);
+}
+
+void rtt_set_counter(uint32_t counter)
+{
+    /* Wait until not busy anymore (should be immediate) */
+    DEBUG("RTT sleeps until safe to write TCNT2\n");
+    _asynch_wait();
+
+    rtt_state.ext_cnt = (uint16_t)(counter >> 8);
+    TCNT2 = (uint8_t)counter;
+}
+
+void rtt_set_alarm(uint32_t alarm, rtt_cb_t cb, void *arg)
+{
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2A);
+    rtt_state.alarm_cb = NULL;
+
+    /* Wait until not busy anymore (should be immediate) */
+    DEBUG("RTT sleeps until safe to write OCR2A\n");
+    _asynch_wait();
+
+    /* Set the alarm value */
+    rtt_state.ext_comp = (uint16_t)(alarm >> 8);
+    OCR2A = (uint8_t)alarm;
+
+    /* Interrupt safe order of assignment */
+    rtt_state.alarm_arg = arg;
+    rtt_state.alarm_cb = cb;
+
+    /* Enable alarm interrupt only if it will trigger before overflow */
+    if (rtt_state.ext_comp <= rtt_state.ext_cnt) {
+        TIMSK2 |= (1 << OCIE2A);
+    }
+}
+
+uint32_t rtt_get_alarm(void)
+{
+    return (((uint32_t)rtt_state.ext_comp << 8) | (uint32_t)OCR2A);
+}
+
+void rtt_clear_alarm(void)
+{
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2A);
+
+    /* Interrupt safe order of assignment */
+    rtt_state.alarm_cb = NULL;
+    rtt_state.alarm_arg = NULL;
+}
+
+void rtt_poweron(void)
+{
+    power_timer2_enable();
+}
+
+void rtt_poweroff(void)
+{
+    power_timer2_disable();
+}
+
+void _asynch_wait(void)
+{
+    /* Wait until all busy flags clear. According to the datasheet,
+     * this can take up to 2 positive edges of TOSC1 (32kHz). */
+    while (ASSR & ((1 << TCN2UB) | (1 << OCR2AUB) | (1 << OCR2BUB)
+                   | (1 << TCR2AUB) | (1 << TCR2BUB))) ;
+}
+
+ISR(TIMER2_OVF_vect) {
+    __enter_isr();
+
+    rtt_state.ext_cnt++;
+
+    /* Enable RTT alarm if overflowed enough times */
+    if (rtt_state.ext_comp == rtt_state.ext_cnt) {
+        TIMSK2 |= (1 << OCIE2A);
```

Comment:
Not clearing will execute the compare match isr imediatly as there is allready a set flag 
/* Clear interrupt flags */
        TIFR2 = (1 << OCF2A);